File: solutions/swift/perfect-numbers/1/Sources/PerfectNumbers/PerfectNumbers.swift
enum Classification { case perfect, abundant, deficient }
enum ClassificationError: Error { case invalidInput }

func classify(number: Int) throws -> Classification {
  guard number > 0 else { throw ClassificationError.invalidInput }

  let aliquot =
    [Int](1..<(number / 2) + 1)
    .filter { number.isMultiple(of: $0) }
    .reduce(0, +)
  return aliquot < number ? .deficient : aliquot > number ? .abundant : .perfect
}

File: solutions/zig/perfect-numbers/3/perfect_numbers.zig
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    var a: usize = 0;
    for (1..n) |f| {
        if (f > n / 2) break;
        if (n % f == 0) a += f;
    }

    if (a < n) return .deficient;
    if (a > n) return .abundant;
    return .perfect;
}
